### PR TITLE
feat!: migrate to ESM

### PR DESCRIPTION
**Pull request by mdonnalley**
### What does this PR do?

Migrate to ESM

### What issues does this PR fix or reference?
@W-14374811@

### CODE REVIEW

File: .lintstagedrc.cjs [hunk 1]
```diff
@@ -0,0 +1,3 @@
+module.exports = {
+  '**/*.{js,json,md}?(x)': () => 'npm run reformat',
+};
```
Comment (cristiand391) [inline .lintstagedrc.cjs:3]: this file can be removed, we don't use lint-staged anymore in husky hooks.

File: package.json [hunk 2]
```diff
@@ -5,59 +5,37 @@
...
-    "@salesforce/dev-config": "^4.0.1",
-    "@salesforce/dev-scripts": "^5.11.0",
+    "@oclif/plugin-command-snapshot": "^5.0.1",
+    "@salesforce/cli-plugins-testkit": "^5",
+    "@salesforce/dev-scripts": "^6.0.1",
     "@salesforce/plugin-command-reference": "^3.0.42",
-    "@salesforce/prettier-config": "^0.0.3",
     "@salesforce/ts-sinon": "1.4.19",
-    "@swc/core": "^1.3.36",
     "@types/inquirer": "^8.2.0",
```
Comment (cristiand391) [inline package.json:20]: there's a v9 of these types avaiable now that it uses inquirer v9.

File: package.json [hunk 3]
```diff
@@ -85,19 +63,20 @@
...
+    },
+    "flexibleTaxonomy": true
   },
   "repository": "salesforcecli/plugin-login",
   "scripts": {
     "build": "wireit",
     "clean": "sf-clean",
     "clean-all": "sf-clean all",
-    "clean:lib": "shx rm -rf lib && shx rm -rf coverage && shx rm -rf .nyc_output && shx rm -f oclif.manifest.json",
+    "clean:lib": "shx rm -rf lib && shx rm -rf coverage && shx rm -rf .nyc_output && shx rm -f oclif.manifest.json oclif.lock",
```
Comment (cristiand391) [inline package.json:74]: nit: we could drop this "clean:lib" from our script and just use `sf-clean` which removes files from a gitignore section:

Screenshot 2023-11-02 at 17 34 40


gitignore template: https://github.com/forcedotcom/dev-scripts/blob/cc3efa96c7b5ae1cce02ac1228187db16bfb230d/files/gitignore#L1-L24

File: package.json [hunk 4]
```diff
@@ -185,7 +163,7 @@
       "output": []
     },
     "test:command-reference": {
-      "command": "\"./bin/dev\" commandreference:generate --erroronwarnings",
+      "command": "ts-node \"./bin/dev.js\" commandreference:generate --erroronwarnings",
```
Comment (cristiand391) [inline package.json:166]: oh, is `ts-node` required for this to work on windows? I was getting a failure on windows on my plugin when moving it to esm:
https://github.com/cristiand391/sf-plugin-api/actions/runs/6714528333/job/18248007631?pr=22

File: src/index.ts [hunk 5]
```diff
@@ -5,4 +5,4 @@
  * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
  */
 
-export = {};
+export default {};
```
Comment (cristiand391) [inline src/index.ts:8]: is this index file still required in esm packages?

**Review by cristiand391 — CHANGES_REQUESTED** [review]
looks good, left some questions/suggestions but there's still a merge conflict to solve

File: src/index.ts (hunk 5 above)
Comment (mdonnalley) [reply src/index.ts:8]: Yes, it's the entry point for oclif/core to figure out the location of the plugin's root directory

File: package.json (hunk 4 above)
Comment (mdonnalley) [reply package.json:166]: Yep, that's correct

File: package.json (hunk 3 above)
Comment (cristiand391) [reply package.json:74]: I logged [W-14423655](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE00001dmkpgYAA/view) to track this.

I'm fine either way, no need to make this change right now 👍🏼

**Review by cristiand391 — APPROVED** [review]

**Comment by cristiand391** [comment]
**QA notes**:

✅ `sf logout` run successfully (can't try logout of any env as I don't have any functions account and our sf plugins don't use those hooks)
✅ `sf login` installs plugin-functions and run login command, opens heroku functions login page in chrome.

**Pull request merged**